File: FE/src/types/teacherUnavailability.ts
export type SessionType = 'MORNING' | 'AFTERNOON'

export type TeacherUnavailabilityStatus = 'PENDING' | 'APPROVED' | 'REJECTED' | 'WITHDRAWN'

export interface TeacherUnavailability {
  id: number
  teacherId: number
  teacherName: string
  semesterId: number
  dayOfWeek?: number | null
  specificDate?: string | null
  validFrom: string
  validTo: string
  session: SessionType
  periodIndexes: string
  note?: string | null
  status: TeacherUnavailabilityStatus
  decisionReason?: string | null
  decidedBy?: number | null
  decidedAt?: string | null
  version: number
  createdAt: string
  updatedAt: string
}

export interface CreateUnavailabilityPayload {
  semesterId: number
  teacherId?: number | null
  dayOfWeek?: number | null
  specificDate?: string | null
  validFrom: string
  validTo: string
  session: SessionType
  periodIndexes: string
  note?: string | null
}

export interface UpdateUnavailabilityPayload {
  expectedVersion: number
  dayOfWeek?: number | null
  specificDate?: string | null
  validFrom: string
  validTo: string
  session: SessionType
  periodIndexes: string
  note?: string | null
}

export interface RejectUnavailabilityPayload {
  expectedVersion: number
  reason: string
}

File: BE/BaiTap-RS/src/main/java/com/JavaTraining/BaiTap_RS/timetable/domain/entity/TeacherUnavailability.java
package com.JavaTraining.BaiTap_RS.timetable.domain.entity;

import java.time.LocalDate;
import java.time.LocalDateTime;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "teacher_unavailability")
@SuppressWarnings("PMD.TooManyFields")
public class TeacherUnavailability {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "unavailability_id", nullable = false)
    private Long id;

    @Column(name = "teacher_id", nullable = false)
    private Long teacherId;

    @Column(name = "semester_id", nullable = false)
    private Long semesterId;

    @Column(name = "day_of_week")
    private Integer dayOfWeek;

    @Column(name = "specific_date")
    private LocalDate specificDate;

    @Column(name = "valid_from", nullable = false)
    private LocalDate validFrom;

    @Column(name = "valid_to", nullable = false)
    private LocalDate validTo;

    @Enumerated(EnumType.STRING)
    @Column(name = "session", nullable = false, length = 20)
    private SessionType session;

    @Column(name = "period_indexes", nullable = false, length = 50)
    private String periodIndexes;

    @Column(name = "note", length = 255)
    private String note;

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private TeacherUnavailabilityStatus status;

    @Column(name = "decision_reason", length = 255)
    private String decisionReason;

    @Column(name = "decided_by")
    private Long decidedBy;

    @Column(name = "decided_at")
    private LocalDateTime decidedAt;

    @Version
    @Column(name = "version", nullable = false)
    private Long version = 0L;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    public TeacherUnavailability(
            Long teacherId,
            Long semesterId,
            Integer dayOfWeek,
            LocalDate specificDate,
            LocalDate validFrom,
            LocalDate validTo,
            SessionType session,
            String periodIndexes,
            String note) {
        this.teacherId = teacherId;
        this.semesterId = semesterId;
        this.dayOfWeek = dayOfWeek;
        this.specificDate = specificDate;
        this.validFrom = validFrom;
        this.validTo = validTo;
        this.session = session;
        this.periodIndexes = periodIndexes;
        this.note = note;
        this.status = TeacherUnavailabilityStatus.PENDING;
    }

    @PrePersist
    protected void onCreate() {
        LocalDateTime now = LocalDateTime.now();
        this.createdAt = now;
        this.updatedAt = now;
        if (this.status == null) {
            this.status = TeacherUnavailabilityStatus.PENDING;
        }
        if (this.version == null) {
            this.version = 0L;
        }
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
